src/index.ts

```diff
@@ -476,138 +662,234 @@ export class Upload extends Pumpify {
+      this.responseHandler(resp);
+    });
+    let headers: GaxiosOptions['headers'] = {};
+
+    // If using multiple chunk upload, set appropriate header
+    if (multiChunkMode && expectedUploadSize) {
+      // The '-1' is because the ending byte is inclusive in the request.
+      const endingByte = expectedUploadSize + this.numBytesWritten - 1;
+      headers = {
+        'Content-Length': expectedUploadSize,
+        'Content-Range': `bytes ${this.offset}-${endingByte}/${this.contentLength}`,
+      };
+    } else {
+      headers = {
+        'Content-Range': `bytes ${this.offset}-*/${this.contentLength}`,
+      };
+    }
+
     const reqOpts: GaxiosOptions = {
       method: 'PUT',
       url: this.uri,
-      headers: {
-        'Content-Range': 'bytes ' + this.offset + '-*/' + this.contentLength,
-      },
-      body: requestStreamEmbeddedStream,
+      headers,
+      body: requestStream,
     };
 
     try {
       await this.makeRequestStream(reqOpts);
-    } catch (e) {
+    } catch (err) {
+      const e = err as Error;
       this.destroy(e);
     }
   }
 
-  private onChunk(
-    chunk: string,
-    enc: BufferEncoding,
-    next: (err?: Error, data?: string) => void
-  ) {
-    const offset = this.offset!;
-    const numBytesWritten = this.numBytesWritten;
+  // Process the API response to look for errors that came in
+  // the response body.
+  private responseHandler(resp: GaxiosResponse) {
+    if (resp.data.error) {
+      this.destroy(resp.data.error);
+      return;
+    }
 
-    this.emit('progress', {
-      bytesWritten: this.numBytesWritten,
-      contentLength: this.contentLength,
-    });
+    const shouldContinueWithNextMultiChunkRequest =
+      this.chunkSize &&
+      resp.status === RESUMABLE_INCOMPLETE_STATUS_CODE &&
+      resp.headers.range;
+
+    if (shouldContinueWithNextMultiChunkRequest) {
+      // Use the upper value in this header to determine where to start the next chunk.
+      // We should not assume that the server received all bytes sent in the request.
+      // https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+      const range: string = resp.headers.range;
+      this.offset = Number(range.split('-')[1]) + 1;
+
+      // We should not assume that the server received all bytes sent in the request.
+      // - https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+      const missingBytes = this.numBytesWritten - this.offset;
+      if (missingBytes) {
+        const dataToPrependForResending = this.lastChunkSent.slice(
+          -missingBytes
+        );
+        // As multi-chunk uploads send one chunk per request and pulls one
+        // chunk into the pipeline, prepending the missing bytes back should
+        // be fine for the next request.
+        this.unshiftChunkBuffer(dataToPrependForResending);
+        this.numBytesWritten -= missingBytes;
+        this.lastChunkSent = Buffer.alloc(0);
```

Comment:
Is the buffer re-allocated for each chunk?

Reply:
Yes, from upstream.
`Buffer.alloc(0)` creates a "Fast" buffer: https://github.com/nodejs/node/blob/e4aa575b05fcf560c6b77b0fa5036328a3417442/lib/buffer.js#L359-L366

Reply:
Okay-- just asking because I know in some languages repeated buffer allocs can be painful. Is there anyway to create one buffer and reuse it for each chunk? Or is that not really how it works in node?

Reply:
Understandable - it doesn't really work that way in Node.js (at least for `Buffer.alloc(0)`)